### PR TITLE
Better http responses

### PR DESCRIPTION
**Pull request by nbrownus**
Trying to get more meaningful http responses back for some operations

### CODE REVIEW

File: http/handler.go [hunk 1]
```diff
@@ -168,14 +168,29 @@ func respondError(w http.ResponseWriter, status int, err error) {
...
+func respondCommon(w http.ResponseWriter, resp *logical.Response, err error) bool {
 	if resp == nil {
 		return false
 	}
 
 	if resp.IsError() {
+		var statusCode int
+
+		switch err {
+		case logical.ErrPermissionDenied:
```
Comment (sethvargo) [inline http/handler.go]: I think we have to be careful/conscious here. Returning a permission denied could potentially verify the existence of a secret. Just like GitHub returns a 404 when you try to access a private repo that you don't have access to, I think we might need to always return a 404.

**Comment by nbrownus** [comment]
This PR doesn't change the error message, just sets the status code to something meaningful. In other words, it only better defines what is already exposed by the underlying apis.

**Comment by sethvargo** [comment]
@nbrownus correct, but now we are exposing that to the user via the HTTP API. Do you see the difference? It's one thing to return an error from the logical backend, but now we are forwarding that error onto the client.

**Comment by nbrownus** [comment]
I'm not following. These errors were already being presented to the clients via the HTTP API, just with a status code of 400 or 500 in every case.

These `curl`s were performed against master

```sh
$ curl -v -H "X-Vault-Token: bad token" http://127.0.0.1:8200/v1/secret/test
> GET /v1/secret/test HTTP/1.1
> User-Agent: curl/7.37.1
> Host: 127.0.0.1:8200
> Accept: */*
> X-Vault-Token: bad token
>
< HTTP/1.1 400 Bad Request
< Content-Type: application/json
< Date: Fri, 19 Jun 2015 21:25:42 GMT
< Content-Length: 33
<
{"errors":["permission denied"]}
```

```sh
$ curl -v http://127.0.0.1:8200/v1/secret/test
> GET /v1/secret/test HTTP/1.1
> User-Agent: curl/7.37.1
> Host: 127.0.0.1:8200
> Accept: */*
>
< HTTP/1.1 400 Bad Request
< Content-Type: application/json
< Date: Fri, 19 Jun 2015 21:25:46 GMT
< Content-Length: 36
<
{"errors":["missing client token"]}
```

These curls against this branch

```sh
$ curl -v -H "X-Vault-Token: bad token" http://127.0.0.1:8200/v1/secret/test
> GET /v1/secret/test HTTP/1.1
> User-Agent: curl/7.37.1
> Host: 127.0.0.1:8200
> Accept: */*
> X-Vault-Token: bad token
>
< HTTP/1.1 403 Forbidden
< Content-Type: application/json
< Date: Fri, 19 Jun 2015 21:41:15 GMT
< Content-Length: 33
<
{"errors":["permission denied"]}
```

```sh
$ curl -v http://127.0.0.1:8200/v1/secret/test
> GET /v1/secret/test HTTP/1.1
> User-Agent: curl/7.37.1
> Host: 127.0.0.1:8200
> Accept: */*
>
< HTTP/1.1 400 Bad Request
< Content-Type: application/json
< Date: Fri, 19 Jun 2015 21:41:31 GMT
< Content-Length: 36
<
{"errors":["missing client token"]}
```

**Comment by sethvargo** [comment]
@nbrownus this is correct. Consider these two responses:

```text
$ curl -v -H "X-Vault-Token: bad token" http://127.0.0.1:8200/v1/secret/test
> GET /v1/secret/test HTTP/1.1
> User-Agent: curl/7.37.1
> Host: 127.0.0.1:8200
> Accept: */*
> X-Vault-Token: bad token
>
< HTTP/1.1 400 Bad Request
< Content-Type: application/json
< Date: Fri, 19 Jun 2015 21:25:42 GMT
< Content-Length: 33
<
{"errors":["permission denied"]}
```

```
$ curl -v -H "X-Vault-Token: bad token" http://127.0.0.1:8200/v1/secret/test
> GET /v1/secret/test HTTP/1.1
> User-Agent: curl/7.37.1
> Host: 127.0.0.1:8200
> Accept: */*
> X-Vault-Token: bad token
>
< HTTP/1.1 403 Forbidden
< Content-Type: application/json
< Date: Fri, 19 Jun 2015 21:41:15 GMT
< Content-Length: 33
<
{"errors":["permission denied"]}
```

This tells an attacker that a secret named `test` exists by returning a 403.

**Comment by nbrownus** [comment]
The response body is `{"errors":["permission denied"]}` in both cases. Changing the response status code to `403` isn't really adding any more information, just making it easier for computers to reason about.

**Comment by armon** [comment]
Thanks! I agree, that this does not expose any additional information

**Comment by kgutwin** [comment]
Note that the website docs (e.g. https://github.com/hashicorp/vault/blob/master/website/source/docs/http/index.html.md#http-status-codes) need to be updated to match.

**Pull request merged**